### PR TITLE
create_protocol(): promote common dosing parameters to plain arguments

### PR DESCRIPTION
**Pull request by Felixmil**
Closes #157

# Build summary: promote Simple-Protocol dosing settings to plain `create_protocol()` arguments

## Implementation

`create_protocol()` previously expressed the core Simple-Protocol dosing settings (dose, start time, end time) only through the free-form `parameters` list, where each `name` had to exactly match a PK-Sim internal parameter name. This change promotes those three settings to plain, unit-aware arguments, mirroring the existing pattern in `create_output_interval()` and `create_event_selection()`, while keeping the `parameters` list as a fully backward-compatible escape hatch.

### Shared dose helper (`R/utils-create.R`)

Two internal, comment-documented helpers were added (neither is exported, so `NAMESPACE` is unchanged):

- `dose_family_units()` returns every unit accepted for a promoted dose, read live from ospsuite as the union of the four real dose-family dimensions: `"Mass"`, `"Amount"`, `"Dose per body weight"`, and `"Dose per body surface area"`. It uses the same tolerant `getUnitsForDimension()` -> `ospUnits[[dim]]` fallback that `validate_unit()` uses, so the accepted set tracks the installed ospsuite and is never hardcoded. This is the single source of truth for both the membership test and the error message. It resolves the spec's repo-grounded correction that there is no `"Input dose"` dimension in ospsuite (verified live: `getUnitsForDimension("Input dose")` throws).
- `resolve_input_dose_parameter(dose, unit, call = parent.frame())` validates `unit` against `dose_family_units()`, aborts with `call = call` (attributing the error to the calling factory) on an invalid unit, and returns the single raw `list(Name = "InputDose", Value = dose, Unit = unit)`. It always emits `Name = "InputDose"`, never a per-body-weight or per-body-surface-area name, and does not validate the numeric `dose` value (the finite-scalar shape check is the caller's responsibility). This matches the contract the sibling schema-item dose work expects, so both factories can share it unchanged.

### Factory wiring (`R/create_protocol.R`)

- Added five arguments to the Simple-Protocol branch: `dose = NULL`, `dose_unit = "mg"`, `start_time = NULL`, `start_time_unit = "h"`, `end_time = NULL`.
- Folded the three value arguments (`dose`, `start_time`, `end_time`) into the existing `schemas` mutual-exclusivity check. The unit arguments carry non-NULL defaults and are deliberately excluded, so a plain `schemas = ...` Advanced call is never spuriously rejected.
- Validate each supplied numeric scalar as a single finite numeric value, using the same wording as the mirrored factories (`{.arg <name>} must be a single finite numeric value`), rejecting `NA`, length-2 vectors, `Inf`, and character input.
- The `End time` parameter's display unit derives from `time_unit`, falling back to `"h"` when `time_unit` is unset, without setting `data$TimeUnit` as a side effect; `data$TimeUnit` remains driven solely by `time_unit`. `time_unit` validation was moved ahead of the branch (validated once) so the derived `End time` unit is never built from an unvalidated `time_unit`.
- Conflict detection between a promoted argument and a same-named `parameters` entry runs against the `Name`-normalised identifiers (via `to_raw_parameters(parameters, "Name")`), so both `create_parameter(name = ...)` and `create_parameter(path = ...)` / raw `list(Path = ...)` entries are caught. Comparison is exact and case-sensitive against the canonical names `InputDose` / `Start time` / `End time`.
- The auto-built entries are placed before user `parameters` in the combined array, and `data$Parameters` is set only when the combined list is non-empty, so a call with no promoted argument and no `parameters` produces byte-identical `data` to before this change.
- Added roxygen `@param` entries for the five new arguments and a runnable promoted-argument example alongside the existing `parameters`-list example; `man/create_protocol.Rd` was regenerated via `devtools::document()` (`NAMESPACE` unchanged).

### Documentation

- `NEWS.md` gains one development-version bullet (placed first, alphabetically before the existing `fraction_unbound()` bullet) describing the new arguments as a validated alternative to hand-building the `InputDose` / `Start time` / `End time` entries, linking `(#157)`.

## Verification

- `NOT_CRAN=true Rscript -e 'devtools::test(filter = "create_protocol")'`: FAIL 0, WARN 0, SKIP 0, PASS 87. The first run recorded the new snapshots in `tests/testthat/_snaps/create_protocol.md`; the recorded error wording was reviewed (invalid dose unit with the live-read accepted units including the UTF-8 units `µg`, `mg/m²`; the finite-scalar messages; the conflict messages including the Path-keyed conflict producing the identical error; the schemas exclusivity naming the promoted arguments) and the re-run was clean.
- `NOT_CRAN=true Rscript -e 'devtools::test()'` (full suite): FAIL 0, PASS 3116. The single warning is a pre-existing, network-dependent `osp_models()` templates-repository test in `test-snapshot.R`, unrelated to this change.
- `Rscript -e 'devtools::document()'`: regenerated `man/create_protocol.Rd`; `NAMESPACE` unchanged (the helpers are internal). Pre-existing unrelated `Snapshot` topic documentation warnings were present before this change and are not affected by it.

New tests cover: the promoted single-dose call producing the expected `InputDose` / `Start time` / `End time` parameters (snapshot); each dose-family unit (`mg`, `mmol`, `mg/kg`, `mg/m²`) yielding exactly one `InputDose` and no per-family parameter; invalid `dose_unit` (`"banana"` and the valid-but-wrong-dimension `"h"`); non-finite / non-scalar / character shape errors; each conflict case plus a Path-keyed and a raw-`Path`-list conflict; `end_time` unit resolution both with `time_unit = "min"` (unit `min`, `data$TimeUnit == "min"`) and without `time_unit` (unit `h`, `data$TimeUnit` unset); `schemas` mutual-exclusivity for each promoted argument plus a bare `schemas = ...` call succeeding with default units; and an explicit backward-compatibility check that the promoted form equals the hand-built `data` and a bare call leaves `data$Parameters` unset.

## Backward compatibility

Confirmed. A call with no new arguments produces byte-identical `data` to before: `data$Parameters` is gated on the combined list being non-empty, so a bare `create_protocol(name = "P")` leaves `data$Parameters` unset, and a `parameters`-only call yields the same `to_raw_parameters(parameters, "Name")` output in the same order. The existing `create_protocol` tests (the simple-protocol and advanced/schemas round-trip tests) stay green unchanged.

## Micro-decisions (auto mode)

All adopted from the plan's locked decisions:

- The start-time unit argument is named `start_time_unit` (default `"h"`), for symmetry with `dose_unit`.
- When `end_time` is supplied and `time_unit` is `NULL`, the `End time` parameter's unit falls back to `"h"` and `data$TimeUnit` stays unset.
- A lone `dose_unit` / `start_time_unit` (value `NULL`) is inert: not validated, not emitted, not an error.
- Only the value arguments participate in the `schemas` mutual-exclusivity check; default unit arguments do not.
- Auto-built parameters precede user `parameters`; conflict detection is exact, case-sensitive, and `Name`-normalised.
- Accepted dose units are read live from ospsuite (never hardcoded).

The plan's Step 1 was executed (the shared helper did not already exist on the branch, as expected). The helper matches the contract the sibling schema-item dose work will reuse unchanged.

# QA: Approved

The PR promotes `dose`/`dose_unit`, `start_time`/`start_time_unit`, and `end_time` to plain arguments on `create_protocol()` exactly as the spec and plan require: a single `InputDose` parameter for any dose-family unit, a shared `resolve_input_dose_parameter()` helper in `R/utils-create.R` validated across the four real ospsuite dimensions (no `"Input dose"` dimension), `Name`-normalised conflict detection catching both Name- and Path-keyed entries, `schemas` mutual-exclusivity, byte-identical no-argument output, regenerated `man/`, a correctly-styled NEWS bullet, and tests covering every enumerated case. All nine acceptance criteria are met and repo conventions hold.

<details>
<summary>Full QA report</summary>

## Method

Read the authoritative `spec.md` (AC-1..AC-9, edge cases, locked decisions) and `plan.md`, then inspected the PR via `gh pr diff 174` / `gh pr view 174`, read the changed source in the read-only worktree (`R/create_protocol.R`, `R/utils-create.R`), the regenerated `man/create_protocol.Rd`, `NEWS.md`, the new tests, and the recorded snapshot file. Verified the commit metadata with `git log origin/main..HEAD`. Tests were not executed locally (R execution out of scope); relied on reading the diff, the test file, and the recorded snapshot text, plus the PR body's reported results (target file: PASS 87, FAIL 0; full suite: PASS 3116, FAIL 0, one pre-existing unrelated network warning).

## AC-by-AC verification

| AC | Verdict | Evidence |
|----|---------|----------|
| AC-1 | Pass | `R/create_protocol.R:126-130` adds `dose = NULL`, `dose_unit = "mg"`, `start_time = NULL`, `start_time_unit = "h"`, `end_time = NULL`. `dose_unit` validated by the dose-family helper (`resolve_input_dose_parameter()` call, line 238). `start_time_unit` validated via `validate_unit(start_time_unit, "Time")` (line 242). `end_time` unit derives from `time_unit` with NULL -> `"h"` fallback (line 229). |
| AC-2 | Pass | `resolve_input_dose_parameter()` (`R/utils-create.R:150-162`) always returns `list(Name = "InputDose", Value = dose, Unit = unit)`, never a per-family name. Test `emits a single InputDose for any dose-family unit` iterates `mg`, `mmol`, `mg/kg`, `mg/m²` and asserts exactly one `InputDose` and zero `Dose`/`DosePerBodyWeight`/`DosePerBodySurfaceArea` entries (`expect_length(input_dose, 1)`, `expect_length(other_dose, 0)`). |
| AC-3 | Pass | `dose_family_units()` (`R/utils-create.R:139-155`) reads the union of the real dimensions `"Mass"`, `"Amount"`, `"Dose per body weight"`, `"Dose per body surface area"` live from ospsuite using the tolerant `getUnitsForDimension()` -> `ospUnits[[dim]]` fallback; it never calls `validate_unit(unit, "Input dose")`. Confirmed no `"Input dose"` string anywhere in the diff. Helper lives in `R/utils-create.R` (the shared location #168 reuses), non-exported, comment-documented, with the `resolve_input_dose_parameter(dose, unit, call = parent.frame())` signature the sibling contract expects. |
| AC-4 | Pass | Conflict detection (`R/create_protocol.R:263-287`) computes `existing_names` from `to_raw_parameters(parameters, "Name")`, which copies `Path` to `Name` when only `Path` is present (`R/utils-create.R:114-118`), so both keyings are caught. Maps `dose`->`InputDose`, `start_time`->`Start time`, `end_time`->`End time`. Tests cover all three Name-keyed conflicts plus a `create_parameter(path = "InputDose")` and a raw `list(Path = "InputDose", ...)` conflict; the recorded snapshot shows the identical error fires for the Path-keyed cases. The `%||% NA_character_` guard safely handles entries lacking both keys. |
| AC-5 | Pass | No `infusion_time` or `volume_of_water_per_body_weight` argument added to the signature; they remain reachable only through `parameters`. Confirmed by the full signature at `R/create_protocol.R:120-134`. |
| AC-6 | Pass | `data$Parameters` is set only when `length(combined) > 0` (`R/create_protocol.R:299-302`), auto entries precede user entries, and `to_raw_parameters(parameters, "Name")` is run only over user `parameters`. Test `leaves output unchanged when no promoted argument is supplied` asserts a `parameters`-only build equals the hand-built `data` list and a bare `create_protocol(name = "P")` leaves `data$Parameters` NULL. Existing suite reported green. |
| AC-7 | Pass | `dose`, `start_time`, `end_time` (value args only) folded into `simple_fields` (`R/create_protocol.R:142-144`), keyed on `is.null`; `dose_unit`/`start_time_unit` (non-NULL defaults) deliberately excluded so a bare `schemas = ...` call does not trip. Tests snapshot the exclusivity error for each of the three value args and assert `expect_s3_class(create_protocol(name = "P", schemas = schemas), "Protocol")` for the bare-schemas success. The `i` hint still lists only the original five Simple fields; this is cosmetic and matches the plan's explicit "no wording change needed" decision, while the `x` line correctly names the conflicting argument. |
| AC-8 | Pass | `NEWS.md:3` adds one dev-version bullet, placed first (sorts before `fraction_unbound()`), single line, function name early, backticked identifiers, `(#157)` before the final period, no internal-helper narration. Roxygen `@param` entries added for all five args (wrapped near 80 cols) plus a runnable promoted-argument example. `man/create_protocol.Rd` regenerated (roxygen-style `\verb{}` for multi-word names, `\code{}` for single tokens, usage block updated); `NAMESPACE` unchanged (helpers internal). No hand-edit signature. |
| AC-9 | Pass | New tests cover: promoted single-dose call (snapshot of `data$Parameters` showing the three canonical entries); each dose-family unit -> single `InputDose`; invalid `dose_unit` (`"banana"` and wrong-dimension `"h"`, snapshot lists accepted units); non-finite/non-scalar/character shape errors; each conflict incl Path-keyed and raw-Path-list; `end_time` unit with `time_unit = "min"` (unit `min`, `data$TimeUnit == "min"`) and without (`unit == "h"`, `expect_null(data$TimeUnit)`); `schemas` exclusivity for each arg plus the bare-schemas success. Uses self-contained `test_that()` blocks, `expect_snapshot(error = TRUE)` for error wording, specific expectations (`expect_length`, `expect_equal`, `expect_s3_class`, `expect_null`). |

## Adversarial / edge-case findings

- Empty `parameters = list()` alongside a promoted arg: `vapply(to_raw_parameters(...), ..., character(1))` returns `character(0)`, `%in%` is FALSE, no spurious conflict. Safe.
- Raw entry lacking both `Name` and `Path`: `param$Name %||% NA_character_` yields `NA_character_`, no false-positive conflict. Safe.
- `time_unit` validation ordering: moved to `R/create_protocol.R:161-163` (before the Simple branch consumes it to build `end_time_unit`) and dropped from the trailing block, so an invalid `time_unit` errors regardless of `end_time`, with no double-abort. Matches plan Step 2e / risk 2.
- `end_time` NULL-`time_unit` side effect: `end_time_unit` is a local; `data$TimeUnit` stays driven solely by the trailing `if (!is.null(time_unit))` block (line 305-307). Edge case 6 satisfied; test asserts `expect_null(data$TimeUnit)`.
- Lone `dose_unit`/`start_time_unit` without their value: unit validation is guarded on the value being non-NULL (dose unit checked only inside the `!is.null(dose)` branch via the helper; `validate_unit(start_time_unit, ...)` only inside `!is.null(start_time)`), so a lone unit is inert, not an error. Matches Decision 3 / edge case 3.
- UTF-8 dose units (`µg`, `mg/m²`, `mg/cm²`) are read live from ospsuite (not hardcoded) and appear correctly encoded in the recorded snapshot's accepted-units line. Matches risk 6.
- Infusion time / Volume of water via `parameters` still pass through: no promotion, no conflict mapping for them, so they flow through `to_raw_parameters()` unchanged. Edge case 11 satisfied.
- `%||%` operator: used throughout the package (rlang is in `Imports`), so the new usage at `R/create_protocol.R:266` resolves.

## Convention compliance

- Base pipe / no magrittr: no pipe introduced; no `%>%`.
- Exported-before-internal ordering: the two new helpers are internal and appended to `R/utils-create.R` alongside the other internal helpers; `create_protocol()` (exported) remains the sole definition in its file. OK.
- Roxygen 80-col wrap: new `@param` blocks wrap at ~80 cols.
- NEWS bullet style: single line, function-name-first, alphabetical placement, backticked, `(#157)` before period. OK.
- `man/` regenerated, not hand-edited: usage block and `\item` entries match roxygen output; `NAMESPACE` unchanged.
- Commit trailer: no `Co-Authored-By` line (verified via `git log origin/main..HEAD`). Commit body uses per-file bullets, imperative summary.
- No dash-as-punctuation in generated prose (NEWS bullet, commit body, PR body use only hyphenated compounds like "Simple-Protocol", "self-documenting", which are allowed).

## Conclusion

All acceptance criteria AC-1..AC-9 are satisfied with direct evidence in the diff, tests, and snapshot. Repo conventions hold. No defects found. Approved.

</details>

## Summary by CodeRabbit

* **New Features**
  * `create_protocol()` now adds promoted Simple-Protocol arguments: `dose`/`dose_unit`, `start_time`/`start_time_unit`, and `end_time`.
  * These are automatically translated into the corresponding Simple-Protocol fields, merged with any extra parameter entries, and conflicts with duplicate entries are rejected.

* **Bug Fixes**
  * Improved validation for promoted dose/unit and time inputs, including clearer error messages and consistent end-time display units (derived from `time_unit`, defaulting to hours).

* **Documentation / Tests**
  * Updated `create_protocol()` documentation and examples; refreshed snapshots and added unit/conflict validation tests.

### CODE REVIEW

File: R/create_protocol.R [hunk 1]
```diff
@@ -24,9 +24,35 @@
...
+#'   `time_unit` is unset; there is no separate end-time unit argument.
+#'   Mutually exclusive with `schemas`, and with an `End time` entry in
+#'   `parameters`.
 #' @param parameters List of [Parameter] objects (created with
 #'   [create_parameter()]) or raw parameter lists for Simple Protocol
-#'   parameters such as start time, end time, and dose.
+#'   parameters such as start time, end time, and dose. The common dose,
+#'   start time, and end time settings can instead be supplied via the
+#'   `dose`, `start_time`, and `end_time` arguments; supplying both a
+#'   promoted argument and the same-named `parameters` entry is an error.
```
Comment (Felixmil) [inline R/create_protocol.R]: Confusion: shoud start time, end time and dose be provided here or not ? this is the most confusing part:  "parameters such as start time, end time, and dose."

File: R/create_protocol.R [hunk 2]
```diff
@@ -24,9 +24,35 @@
...
+#'   mass, amount, dose per body weight, or dose per body surface area
+#'   dimensions, and the unit alone selects the family (the emitted
+#'   parameter is always a single `InputDose`). Only consulted when `dose`
+#'   is supplied.
+#' @param start_time Numeric scalar. Optional start time for a Simple
+#'   Protocol. When supplied it is emitted as a `Start time` parameter.
+#'   Mutually exclusive with `schemas`, and with a `Start time` entry in
+#'   `parameters`.
+#' @param start_time_unit Character. Display unit for `start_time`,
+#'   default `"h"`, validated against dimension `"Time"`. Only consulted
```
Comment (Felixmil) [inline R/create_protocol.R]: ospsuite::ospUnits$Time should be explicitely mentioned here.

similarly for other parameters: show supported units by referencing the right ospsuite::... calls.

File: R/create_protocol.R [hunk 3]
```diff
@@ -24,9 +24,35 @@
...
+#'   supplied it is emitted as a single `InputDose` parameter carrying
+#'   `dose` and `dose_unit`. Mutually exclusive with `schemas`, and with
+#'   an `InputDose` entry in `parameters`.
+#' @param dose_unit Character. Display unit for `dose`, default `"mg"`.
+#'   Validated as a dose-family unit: the unit must belong to one of the
+#'   mass, amount, dose per body weight, or dose per body surface area
+#'   dimensions, and the unit alone selects the family (the emitted
+#'   parameter is always a single `InputDose`). Only consulted when `dose`
+#'   is supplied.
+#' @param start_time Numeric scalar. Optional start time for a Simple
```
Comment (Felixmil) [inline R/create_protocol.R:41]: instead of _time and _time_unit, use time() helper ? It would take care of validation so some validation in the function could be removed.

File: R/create_protocol.R [hunk 4]
```diff
@@ -24,9 +24,35 @@
 #' @param target_organ Character. Target organ for the dose.
 #' @param target_compartment Character. Target compartment for the
 #'   dose.
+#' @param dose Numeric scalar. Optional dose for a Simple Protocol. When
```
Comment (Felixmil) [inline R/create_protocol.R:27]: create a dose() helper ?

Comment (Felixmil) [reply R/create_protocol.R:27]: Declined for this PR to keep the paired `value` / `unit` convention that issue #157 (and the promotion-scoping work in #158) specified, and that the companion promotion issues #167, #168, and #169 all follow. Introducing a `dose()` value-object constructor is a batch-wide API decision rather than something to settle inside one factory, so I captured it as a design spike in #177.

File: R/create_protocol.R (hunk 3 above)
Comment (Felixmil) [reply R/create_protocol.R:41]: Declined here for the same reason as the `dose()` suggestion: this PR keeps the paired `value` / `unit` convention shared across the promotion issues #157, #167, #168, and #169. There is an added wrinkle for time specifically: `time()` is already exported, but as a vector time-series constructor for `create_observed_data()` (`R/value_specs.R`), so it takes a numeric vector rather than a scalar. Reusing that name for a scalar dosing-time value object, or adding a second time helper, is a public-API decision I would rather not make inside one factory. I tracked it in the value-object design spike #177.

File: R/create_protocol.R (hunk 2 above)
Comment (Felixmil) [reply R/create_protocol.R]: Done. The roxygen now cites the ospsuite enums explicitly: `time_unit` and `start_time_unit` point to `ospsuite::ospUnits$Time`. For `dose_unit` there is no single `ospsuite::ospUnits$Dose` member (the resolver unions the dose-family dimensions), so it cites the four backing enums instead: `ospsuite::ospUnits$Mass`, `ospsuite::ospUnits$Amount`, `ospsuite::ospUnits[["Dose per body weight"]]`, and `ospsuite::ospUnits[["Dose per body surface area"]]`.

File: R/create_protocol.R (hunk 1 above)
Comment (Felixmil) [reply R/create_protocol.R]: Done. The `parameters` roxygen now leads with its actual purpose (the free-form escape hatch for any Simple-Protocol parameter that has no dedicated argument) and then states that dose, start time, and end time each have a dedicated `dose` / `start_time` / `end_time` argument, and that supplying one of those settings both ways is an error. The old "parameters such as start time, end time, and dose" phrasing that read as if those belonged here is gone.

File: R/create_protocol.R [hunk 5]
```diff
@@ -87,6 +123,11 @@ create_protocol <- function(
...
   target_organ = NULL,
   target_compartment = NULL,
+  dose = NULL,
+  dose_unit = "mg",
+  start_time = NULL,
+  start_time_unit = "h",
+  end_time = NULL,
   parameters = NULL,
   schemas = NULL,
   time_unit = NULL
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply R/create_protocol.R]: Fixed. The five promoted arguments are now appended after `parameters`, `schemas`, and `time_unit`, so the original positional order is restored. Added a regression test that passes `parameters` in the sixth positional slot.

File: R/create_protocol.R [hunk 6]
```diff
@@ -98,6 +139,9 @@ create_protocol <- function(
       dosing_interval = dosing_interval,
       target_organ = target_organ,
       target_compartment = target_compartment,
+      dose = dose,
+      start_time = start_time,
+      end_time = end_time,
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply R/create_protocol.R:150]: Fixed. The schema-conflict guidance now lists `dose`, `start_time`, and `end_time` alongside the other Simple-Protocol fields. Snapshot updated accordingly.

File: R/utils-create.R [hunk 7]
```diff
@@ -126,6 +126,62 @@ to_raw_parameters <- function(parameters, name_key) {
...
+  if (!(unit %in% valid_units)) {
+    cli::cli_abort(
+      c(
+        "Invalid dose unit: {unit}",
+        "i" = "Valid dose units are: {toString(valid_units)}"
+      ),
+      call = call
+    )
+  }
+  list(Name = "InputDose", Value = dose, Unit = unit)
```
[the remark this answers is not in the document]
Comment (Felixmil) [reply R/utils-create.R:182]: Fixed. `resolve_input_dose_parameter()` now guards `unit` with `is_non_empty_scalar_string()` before the `%in%` membership test, so a `NULL`, empty, or multi-length `dose_unit` routes through the same "Invalid dose unit" abort instead of a raw base-R error. Added tests covering `dose_unit = NULL` and a multi-length `dose_unit`.
